lib/common/uint128.h

```diff
@@ -0,0 +1,273 @@
+    *this += 1;
+    return *this;
+  }
+  constexpr UnsignedInt128 operator++(int /*postfix*/) {
+    UnsignedInt128 result{*this};
+    *this += 1;
+    return result;
+  }
+  constexpr UnsignedInt128 operator--(/*prefix*/) {
+    *this -= 1;
+    return *this;
+  }
+  constexpr UnsignedInt128 operator--(int /*postfix*/) {
+    UnsignedInt128 result{*this};
+    *this -= 1;
+    return result;
+  }
+
+  constexpr UnsignedInt128 operator&(UnsignedInt128 that) const {
+    return {high_ & that.high_, low_ & that.low_};
+  }
+  constexpr UnsignedInt128 operator|(UnsignedInt128 that) const {
+    return {high_ | that.high_, low_ | that.low_};
+  }
+  constexpr UnsignedInt128 operator^(UnsignedInt128 that) const {
+    return {high_ ^ that.high_, low_ ^ that.low_};
+  }
+
+  constexpr UnsignedInt128 operator<<(UnsignedInt128 that) const {
+    if (that >= 128) {
+      return {};
+    } else if (that == 0) {
+      return *this;
+    } else {
+      std::uint64_t n{that.low_};
+      if (n >= 64) {
+        return {low_ << (n - 64), 0};
+      } else {
+        return {(high_ << n) | (low_ >> (64 - n)), low_ << n};
+      }
+    }
+  }
+  constexpr UnsignedInt128 operator>>(UnsignedInt128 that) const {
+    if (that >= 128) {
+      return {};
+    } else if (that == 0) {
+      return *this;
+    } else {
+      std::uint64_t n{that.low_};
+      if (n >= 64) {
+        return {0, high_ >> (n - 64)};
+      } else {
+        return {high_ >> n, (high_ << (64 - n)) | (low_ >> n)};
+      }
+    }
+  }
+
+  constexpr UnsignedInt128 operator+(UnsignedInt128 that) const {
+    std::uint64_t lower{(low_ & ~topBit) + (that.low_ & ~topBit)};
+    bool carry{((lower >> 63) + (low_ >> 63) + (that.low_ >> 63)) > 1};
+    return {high_ + that.high_ + carry, low_ + that.low_};
+  }
+  constexpr UnsignedInt128 operator-(UnsignedInt128 that) const {
+    return *this + -that;
+  }
+
+  constexpr UnsignedInt128 operator*(UnsignedInt128 that) const {
+    std::uint64_t mask32{0xffffffff};
+    if (high_ == 0 && that.high_ == 0) {
+      std::uint64_t x0{low_ & mask32}, x1{low_ >> 32};
+      std::uint64_t y0{that.low_ & mask32}, y1{that.low_ >> 32};
+      UnsignedInt128 x0y0{x0 * y0}, x0y1{x0 * y1};
+      UnsignedInt128 x1y0{x1 * y0}, x1y1{x1 * y1};
+      return x0y0 + ((x0y1 + x1y0) << 32) + (x1y1 << 64);
+    } else {
+      std::uint64_t x0{low_ & mask32}, x1{low_ >> 32}, x2{high_ & mask32},
+          x3{high_ >> 32};
+      std::uint64_t y0{that.low_ & mask32}, y1{that.low_ >> 32},
+          y2{that.high_ & mask32}, y3{that.high_ >> 32};
+      UnsignedInt128 x0y0{x0 * y0}, x0y1{x0 * y1}, x0y2{x0 * y2}, x0y3{x0 * y3};
+      UnsignedInt128 x1y0{x1 * y0}, x1y1{x1 * y1}, x1y2{x1 * y2};
+      UnsignedInt128 x2y0{x2 * y0}, x2y1{x2 * y1};
+      UnsignedInt128 x3y0{x3 * y0};
+      return x0y0 + ((x0y1 + x1y0) << 32) + ((x0y2 + x1y1 + x2y0) << 64) +
+          ((x0y3 + x1y2 + x2y1 + x3y0) << 96);
+    }
+  }
```

Comment:
Should some of these more complex functions be defined out-of-line?

Reply:
They need to be `constexpr` so it doesn't really make a difference.